pkg/sql/sem/builtins/generator_builtins.go

```diff
@@ -1231,6 +1241,167 @@ func (g *jsonEachGenerator) Values() (tree.Datums, error) {
 	return tree.Datums{g.key, g.value}, nil
 }
 
+var jsonPopulateProps = tree.FunctionProperties{
+	Class:    tree.GeneratorClass,
+	Category: categoryGenerator,
+	// The typical way to call json_populate_record is to send NULL::atype as the
+	// first argument, so we have to accept nullable args.
+	NullableArgs: true,
+}
+
+func makeJSONPopulateImpl(gen tree.GeneratorWithExprsFactory, info string) tree.Overload {
+	return tree.Overload{
+		// The json{,b}_populate_record{,set} builtins all have a 2 argument
+		// structure. The first argument is an arbitrary tuple type, which is used
+		// to set the columns of the output when the builtin is used as a FROM
+		// source, or used as-is when it's used as an ordinary projection.
+		// The second argument is a JSON object or array of objects. The builtin
+		// transforms the JSON in the second argument into the tuple in the first
+		// argument, field by field, casting fields in key "k" to the type in the
+		// tuple slot "k". Any tuple fields that were missing in the JSON will be
+		// left as they are in the input argument.
+		// The first argument can be of the form NULL::<tupletype>, in which case
+		// the default values of each field will be NULL.
+		// The second argument can also be null, in which case the first argument
+		// is returned as-is.
+		Types: tree.ArgTypes{{"base", types.Any}, {"from_json", types.Jsonb}},
+		ReturnType: func(args []tree.TypedExpr) *types.T {
+			if len(args) != 2 {
+				return tree.UnknownReturnType
+			}
+			return args[0].ResolvedType()
+		},
+		GeneratorWithExprs: gen,
+		Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+			return nil, newUnsuitableUseOfGeneratorError()
+		},
+		Info:       info,
+		Volatility: tree.VolatilityStable,
+	}
+}
+
+func makeJSONPopulateRecordGenerator(
+	evalCtx *tree.EvalContext, args tree.Exprs,
+) (tree.ValueGenerator, error) {
+	evalled := make(tree.Datums, len(args))
+	for i := range args {
+		var err error
+		evalled[i], err = args[i].(tree.TypedExpr).Eval(evalCtx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	tupleType := args[0].(tree.TypedExpr).ResolvedType()
+	var defaultElems tree.Datums
+	if evalled[0] == tree.DNull {
+		defaultElems = make(tree.Datums, len(tupleType.TupleLabels()))
+		for i := range defaultElems {
+			defaultElems[i] = tree.DNull
+		}
+	} else {
+		defaultElems = tree.MustBeDTuple(evalled[0]).D
+	}
+	d := evalled[1]
+	if evalled[1] != tree.DNull {
+		j := tree.MustBeDJSON(evalled[1])
+		if j.Type() != json.ObjectJSONType {
+			return nil, pgerror.Newf(pgcode.InvalidParameterValue, "argument of json_populate_record must be an object")
+		}
+		builder := json.ArrayBuilder{}
+		builder.Add(j.JSON)
+		jarr := builder.Build()
+		d = tree.NewDJSON(jarr)
```

Comment:
I think it'd be simpler to make two generator structs that implement a ValuesGenerator, one for json_populate_record and one for json_populate_recordset. The generator for json_populate_record would be very simple and you wouldn't have to synthesize a JSON array like this.